main.go

```diff
@@ -453,25 +505,220 @@ func (c *controller) waitForCacheSync(stop <-chan struct{}) error {
 	return nil
 }
 
-func (c *controller) worker() {
-	for c.queue.get() {
+func (c *controller) podWorker() {
+	fn := func() bool {
+		key, quit := c.podQ.Get()
+		if quit {
+			return false
+		}
+		defer c.podQ.Done(key)
+		if err := c.cleanUp(key.(*corev1.Pod)); err != nil {
+			level.Error(c.logger).Log("msg", "unable to clean up PVC", "err", err)
+			c.podQ.AddRateLimited(key)
+			return true
+		}
+		c.podQ.Forget(key)
+		return true
+	}
+	for fn() {
+	}
+}
+
+func (c *controller) stsWorker() {
+	for c.stsQ.get() {
 		c.sync()
 	}
 }
 
+func (c *controller) resolvePodOwnerRef(namespace string, refs []metav1.OwnerReference) (*appsv1.StatefulSet, error) {
+	for _, ref := range refs {
+		// If the owner reference points at the wrong kind of object, skip.
+		if ref.Kind != "StatefulSet" {
+			continue
+		}
+		// If the owner reference points at something that we don't have, then skip.
+		obj, ok, err := c.ssetInf.GetStore().GetByKey(fmt.Sprintf("%s/%s", namespace, ref.Name))
+		if !ok {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		sts := obj.(*appsv1.StatefulSet)
+		if sts.UID != ref.UID {
+			return nil, errors.Wrap(err, "owner reference UID does not match StatefulSet")
+		}
+		return sts, nil
+	}
+	return nil, nil
+}
+
+func (c *controller) generateHelper(name string, sts *appsv1.StatefulSet) *corev1.Pod {
+	initContainerTemplate := corev1.Container{
```

Comment:
is this the right pattern?

Reply:
Is there some established way to execute multiple containers serially? I would try to stay away from shell scripting as you have so far, but this seems a little hacky.

Reply:
init containers by design all execute in series; the `rm -rf` container could even be a second initcontainer. so I don't think this is much of an abuse. another option is to have multiple jobs that are dispatched one after another, i.e. one backup job, then a cleanup job. However, this seems like a lot of book-keeping for not much benefit

Reply:
Yeah agreed. Let’s go with this strategy.